homeassistant/components/huawei_lte/device_tracker.py

```diff
@@ -1,63 +1,189 @@
+    SOURCE_TYPE_ROUTER,
+)
+from homeassistant.components.device_tracker.config_entry import ScannerEntity
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import CONF_URL
-from . import RouterData
-from .const import DOMAIN, KEY_WLAN_HOST_LIST
+from homeassistant.helpers import entity_registry
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from . import HuaweiLteBaseEntity
+from .const import DEFAULT_TRACK_NEW, DOMAIN, KEY_WLAN_HOST_LIST, UPDATE_SIGNAL
 
 
 _LOGGER = logging.getLogger(__name__)
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({vol.Optional(CONF_URL): cv.url})
-
-HOSTS_PATH = f"{KEY_WLAN_HOST_LIST}.Hosts.Host"
-
-
-def get_scanner(hass, config):
-    """Get a Huawei LTE router scanner."""
-    data = hass.data[DOMAIN].get_data(config)
-    data.subscribe(HOSTS_PATH)
-    return HuaweiLteScanner(data)
+_NEW_DEVICE_SCAN = f"{DEVICE_TRACKER_DOMAIN}/new_device_scan"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up from config entry."""
+
+    # Grab hosts list once to examine whether the initial fetch has got some data for
+    # us, i.e. if wlan host list is supported. Only set up a subscription and proceed
+    # with adding and tracking entities if it is.
+    router = hass.data[DOMAIN].routers[config_entry.data[CONF_URL]]
+    try:
+        _ = router.data[KEY_WLAN_HOST_LIST]["Hosts"]["Host"]
+    except KeyError:
+        _LOGGER.debug("%s[%s][%s] not in data", KEY_WLAN_HOST_LIST, "Hosts", "Host")
+        return
+
+    # Initialize already tracked entities
+    tracked: Set[str] = set()
+    registry = await entity_registry.async_get_registry(hass)
+    for entity in registry.entities.values():
+        if (
+            entity.domain == DEVICE_TRACKER_DOMAIN
+            and entity.config_entry_id == config_entry.entry_id
+        ):
+            tracked.add(entity.unique_id)
+    await async_add_new_entities(hass, router.url, async_add_entities, tracked, True)
+
+    # Tell parent router to poll hosts list for new devices, if enabled
+    if config_entry.options.setdefault(CONF_TRACK_NEW, DEFAULT_TRACK_NEW):
+        router.subscriptions[KEY_WLAN_HOST_LIST].add(_NEW_DEVICE_SCAN)
+
+    async def _async_maybe_add_new_entities(url: str) -> None:
+        """Add new entities if enabled and the update signal comes from our router."""
+        if url == router.url and config_entry.options.get(CONF_TRACK_NEW):
+            await async_add_new_entities(hass, url, async_add_entities, tracked)
+
+    # Register to handle router data updates
+    disconnect_dispatcher = async_dispatcher_connect(
+        hass, UPDATE_SIGNAL, _async_maybe_add_new_entities
+    )
+    router.unload_handlers.append(disconnect_dispatcher)
+
+    # Add new entities from initial scan, if enabled
+    if config_entry.options.get(CONF_TRACK_NEW):
+        await async_add_new_entities(hass, router.url, async_add_entities, tracked)
+
+
+async def async_add_new_entities(
```

Comment:
This should be a callback instead of a coroutine since we don't await inside. Callbacks are a bit faster than coroutines.

Reply:
You mean just removing the async (and renaming as appropriate), i.e. turning it to a regular function?

Reply:
It's still good to name it `async_add_new_entities` so it's clear that it should run in the event loop. Just remove `async ` in `async def` and don't await it.

Reply:
Azure confuses me. Tests failed with an apparently unrelated error (plex tests timeouting). When I click re-run, either nothing happens, or the test I tried to re-run gets silently immediately marked as cancelled without any explanation.

Oh well, maybe I'll just merge dev here again to get it to re-run.